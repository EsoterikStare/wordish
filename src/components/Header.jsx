import React from 'react';
import { string } from 'prop-types';
import {
  Button,
  Grid,
  Typography,
  styled,
} from '@mui/material';
import { Settings as SettingsIcon, Share2 } from 'react-feather';

import StatsDialog from './StatsDialog';
import Settings from './Settings';

import { useAppState } from '../AppStateContext';

const HeaderButton = styled(Button)(() => ({
  backgroundColor: '#333',
  color: 'white',
  height: 40,
  minWidth: 40,
  padding: 4,
  userSelect: 'none',
  '&:hover': {
    backgroundColor: '#333',
  },
}));

const Header = ({ title }) => {
  const { state } = useAppState();
  const actionKeys = [' ', 'Enter'];

  const [dialogOpen, setDialogOpen] = React.useState(false);
  const [settingsAnchor, setSettingsAnchor] = React.useState();
  const onSettingsClick = (e) => {
    setSettingsAnchor(e.target);
  };
  const onSettingsClose = () => setSettingsAnchor(undefined);

  React.useEffect(() => {
    if (state.gameState !== 'playing') {
      setDialogOpen(true);
    }
  }, [state.gameState]);

  return (
    <>
      <Grid id="header-grid" alignItems="center" container justifyContent="center" item spacing={4}>
        <Grid id="stats-dialog-grid" item>
          <HeaderButton
            aria-label="Stats and Share"
            onClick={() => setDialogOpen(true)}
            onKeyDown={(e) => {
              if (actionKeys.includes(e.key)) {
                setDialogOpen(true);
              }
            }}
          >
            <Share2 />
          </HeaderButton>
        </Grid>
        <Grid item xs="auto">
          <Typography
            component="h1"
            sx={{
              color: 'white',
              fontSize: '2.5rem',
              fontWeight: 'bold',
              userSelect: 'none',
            }}
          >
            {title}
          </Typography>
        </Grid>
        <Grid id="settings-button-grid" item>
          <HeaderButton
            aria-label="settings"
            onClick={onSettingsClick}
            onKeyDown={(e) => {
              if (actionKeys.includes(e.key)) {
                onSettingsClick(e);
              }
            }}
          >
            <SettingsIcon style={{ pointerEvents: 'none' }} />
          </HeaderButton>
          <Settings anchorEl={settingsAnchor} onClose={onSettingsClose} />
        </Grid>
      </Grid>
      <StatsDialog open={dialogOpen} onClose={() => setDialogOpen(false)} />
    </>
  );
};

Header.propTypes = {
  title: string.isRequired,
};

export default Header;
