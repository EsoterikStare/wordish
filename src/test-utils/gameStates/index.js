import loseGameStateJSON from './loseGameState.json';
import midGuessGameStateJSON from './midGuessGameState.json';
import newGameStateJSON from './newGameState.json';
import winGameStateJSON from './winGameState.json';

export const loseGameState = JSON.stringify(loseGameStateJSON);
export const midGuessGameState = JSON.stringify(midGuessGameStateJSON);
export const newGameState = JSON.stringify(newGameStateJSON);
export const winGameState = JSON.stringify(winGameStateJSON);
