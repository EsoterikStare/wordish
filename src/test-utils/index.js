export * from './gameStates';
export { default as setup } from './setup';
