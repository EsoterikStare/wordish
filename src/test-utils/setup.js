import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { node } from 'prop-types';
import React from 'react';

import { AppStateProvider } from '../AppStateContext';
import AppThemeProvider from '../AppThemeProvider';

const Wrapper = ({ children }) => (
  <AppStateProvider>
    <AppThemeProvider>{children}</AppThemeProvider>
  </AppStateProvider>
);
Wrapper.propTypes = {
  children: node.isRequired,
};
Wrapper.defaultProps = {};

const setup = jsx => {
  const { rerender, ...rest } = render(<Wrapper>{jsx}</Wrapper>);
  return {
    user: userEvent.setup(),
    rerender: newJsx => rerender(<Wrapper>{newJsx}</Wrapper>),
    ...rest,
  };
};

export default setup;
