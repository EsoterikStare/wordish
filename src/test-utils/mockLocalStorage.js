import { newGameState } from './gameStates';

const mockLocalStorage = (gameState = newGameState) => {
  const originalLocalStorage = { ...window.localStorage };
  const mockValues = {
    clear: jest.fn(),
    // Provide a known win state from localStorage mock that will be used by AppStateProvider.
    getItem: jest.fn(() => gameState),
    removeItem: jest.fn(),
    setItem: jest.fn(),
  };

  // Mock window.localStorage with assertable functions.
  Object.defineProperty(window, 'localStorage', {
    value: mockValues,
    writable: true,
  });

  const resetLocalStorage = () => {
    Object.defineProperty(window, 'localStorage', {
      value: originalLocalStorage,
    });
  };

  return {
    resetLocalStorage,
    ...mockValues,
  };
};

export default mockLocalStorage;
