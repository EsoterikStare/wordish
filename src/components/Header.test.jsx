import { screen } from '@testing-library/react';
import React from 'react';

import Header from './Header';
import { setup } from '../test-utils';

describe('Header', () => {
  it('should be defined', () => {
    expect(Header).toBeDefined();
  });
  it('should render', () => {
    setup(<Header title="WORDISH" />);
    expect(screen.getByTestId('header-grid')).toBeInTheDocument();
  });
  it('should render a heading with the given title prop text', () => {
    const title = 'WORDISH';
    setup(<Header title={title} />);

    expect(screen.getByRole('heading', { name: title })).toBeInTheDocument();
  });
  describe('Stats button', () => {
    it('should render', () => {
      setup(<Header title="WORDISH" />);
      expect(screen.getByRole('button', { name: 'Stats and Share' })).toBeInTheDocument();
    });
    it('should open a dialog when clicked', async () => {
      const { user } = setup(<Header title="WORDISH" />);
      await user.click(screen.getByRole('button', { name: 'Stats and Share' }));

      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
  });
  describe('Settings button', () => {
    it('should render', () => {
      setup(<Header title="WORDISH" />);
      expect(screen.getByRole('button', { name: 'Settings' })).toBeInTheDocument();
    });
    it('should open a menu when clicked', async () => {
      const { user } = setup(<Header title="WORDISH" />);
      await user.click(screen.getByRole('button', { name: 'Settings' }));

      expect(screen.getByRole('menu')).toBeInTheDocument();
    });
  });
});
