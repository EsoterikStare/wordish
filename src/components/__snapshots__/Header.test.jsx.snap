// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should have it's components arranged correctly 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-4 css-1mqp8pp-MuiGrid-root"
  data-testid="header-grid"
  role="banner"
>
  <div
    class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    id="stats-dialog-grid"
  >
    <button
      aria-label="Stats and Share"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-15wrqm4-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="18"
          cy="5"
          r="3"
        />
        <circle
          cx="6"
          cy="12"
          r="3"
        />
        <circle
          cx="18"
          cy="19"
          r="3"
        />
        <line
          x1="8.59"
          x2="15.42"
          y1="13.51"
          y2="17.49"
        />
        <line
          x1="15.41"
          x2="8.59"
          y1="6.51"
          y2="10.49"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-1wrgmsj-MuiGrid-root"
    data-testid="title"
  >
    <h1
      class="MuiTypography-root MuiTypography-body1 css-5hh29c-MuiTypography-root"
    >
      WORDISH
    </h1>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    data-testid="settings-button-grid"
  >
    <button
      aria-label="Settings"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-15wrqm4-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="pointer-events: none;"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="3"
        />
        <path
          d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
